src/Catalyst.Core.Modules.Dfs.Tests/IntegrationTests/DagNodeTest.cs

```diff
@@ -0,0 +1,309 @@
+
+            RoundtripTest(cnode);
+        }
+
+        [Fact]
+        public void RemoveLink()
+        {
+            var a = Encoding.UTF8.GetBytes("a");
+            var anode = new DagNode(a);
+
+            var b = Encoding.UTF8.GetBytes("b");
+            var bnode = new DagNode(b);
+
+            var c = Encoding.UTF8.GetBytes("c");
+            var cnode = new DagNode(c, new[]
+            {
+                anode.ToLink(), bnode.ToLink()
+            });
+
+            var dnode = cnode.RemoveLink(anode.ToLink());
+            Assert.False(ReferenceEquals(dnode, cnode));
+            Assert.Equal(1, dnode.DataBytes.Length);
+            Assert.Equal(1, dnode.Links.Count());
+            Assert.Equal(bnode.Id, (string) dnode.Links.First().Id);
+            Assert.Equal(bnode.Size, dnode.Links.First().Size);
+
+            RoundtripTest(cnode);
+        }
+
+        [Fact]
+        public void NullStream()
+        {
+            TestUtils.ExceptionAssert.Throws(() => new DagNode((CodedInputStream) null));
+            TestUtils.ExceptionAssert.Throws(() => new DagNode((Stream) null));
+        }
+
+        [Fact]
+        public void LinkWithCidv1()
+        {
+            var data =
+                "124F0A4401551340309ECC489C12D6EB4CC40F50C902F2B4D0ED77EE511A7C7A9BCD3CA86D4CD86F989DD35BC5FF499670DA34255B45B0CFD830E81F605DCF7DC5542E93AE9CD76F120568656C6C6F180B0A020801"
+                   .ToHexBuffer();
+            var ms = new MemoryStream(data, false);
+            var node = new DagNode(ms);
+            Assert.Equal("0801", node.DataBytes.ToHexString());
+            Assert.Equal(1, node.Links.Count());
+            var link = node.Links.First();
+            Assert.Equal("hello", link.Name);
+            Assert.Equal(1, link.Id.Version);
+            Assert.Equal("raw", link.Id.ContentType);
+            Assert.Equal("sha2-512", link.Id.Hash.Algorithm.Name);
+            Assert.Equal(11, link.Size);
+        }
+
+        [Fact]
+        public void SettingId()
+        {
+            var a = new DagNode((byte[]) null);
+            var b = new DagNode((byte[]) null)
+            {
+                // Wrong hash but allowed.
+                Id = "QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1m"
+            };
+            Assert.Equal(a.DataBytes.Length, b.DataBytes.Length);
+            Assert.Equal(a.Links.Count(), b.Links.Count());
+            Assert.Equal(a.Size, b.Size);
+            Assert.NotEqual(a.Id, b.Id);
+
+            RoundtripTest(b);
+        }
+
+        private static void RoundtripTest(IDagNode a)
+        {
+            var ms = new MemoryStream();
+            a.Write(ms);
+            ms.Position = 0;
+            var b = new DagNode(ms);
+            
+            Assert.Equal(a.DataBytes, b.DataBytes);
+            Assert.Equal(a.ToArray(), b.ToArray());
+            Assert.Equal(a.Links.Count(), b.Links.Count());
+            foreach (var _ in a.Links.Zip(b.Links, (first, second) =>
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)